plugin/pkg/admission/limitranger/admission_test.go

```diff
@@ -856,3 +858,75 @@ func TestPersistentVolumeClaimLimitFunc(t *testing.T) {
 		}
 	}
 }
+
+// TestLimitRanger_GetLimitRangesFixed22422 Fixed Admission controllers can cause unnecessary significant load on apiserver #22422
+func TestLimitRanger_GetLimitRangesFixed22422(t *testing.T) {
+	limitRange := validLimitRangeNoDefaults()
+	limitRanges := []corev1.LimitRange{limitRange}
+
+	var count int64
+	mockClient := &fake.Clientset{}
+
+	unhold := make(chan struct{})
+	mockClient.AddReactor("list", "limitranges", func(action core.Action) (bool, runtime.Object, error) {
+		atomic.AddInt64(&count, 1)
+
+		limitRangeList := &corev1.LimitRangeList{
+			ListMeta: metav1.ListMeta{
+				ResourceVersion: fmt.Sprintf("%d", len(limitRanges)),
+			},
+		}
+		for index, value := range limitRanges {
+			value.ResourceVersion = fmt.Sprintf("%d", index)
+			limitRangeList.Items = append(limitRangeList.Items, value)
+		}
+		// he always blocking before sending the signal
+		<-unhold
+		return true, limitRangeList, nil
+	})
+
+	handler, _, err := newHandlerForTest(mockClient)
+	if err != nil {
+		t.Errorf("unexpected error initializing handler: %v", err)
+	}
+
+	attributes := admission.NewAttributesRecord(nil, nil, api.Kind("kind").WithVersion("version"), "test", "name", api.Resource("resource").WithVersion("version"), "subresource", admission.Create, &metav1.CreateOptions{}, false, nil)
```

Comment:
does anything test the namespace sharding of the singleflight to make sure that is working as expected?

Reply:
I thought about it, but since you hit the lrucache first, I couldn't see how to properly exercise it

```
		lruItemObj, ok := l.liveLookupCache.Get(a.GetNamespace())
```

Reply:
Is this approach testable?

```
diff --git a/plugin/pkg/admission/limitranger/admission_test.go b/plugin/pkg/admission/limitranger/admission_test.go
index ea3853afecb..10f66b9759c 100644
--- a/plugin/pkg/admission/limitranger/admission_test.go
+++ b/plugin/pkg/admission/limitranger/admission_test.go
@@ -867,7 +867,8 @@ func TestLimitRanger_GetLimitRangesFixed22422(t *testing.T) {
        var count int64
        mockClient := &fake.Clientset{}
 
-       unhold := make(chan struct{})
+       unhold := make(chan struct{}, 1)
+
        mockClient.AddReactor("list", "limitranges", func(action core.Action) (bool, runtime.Object, error) {
                atomic.AddInt64(&count, 1)
 
@@ -908,6 +909,15 @@ func TestLimitRanger_GetLimitRangesFixed22422(t *testing.T) {
        unhold <- struct{}{}
        // and here we wait for all the goroutines
        wg.Wait()
+
+       // he will not block if there is a cache failure
+       unhold <- struct{}{}
+       // This call will use the cache, and if the count changes, it will be an error
+       _, err = handler.GetLimitRanges(attributes)
+       if err != nil {
+               t.Errorf("unexpected error: %v", err)
+       }
+
        // since all the calls were holded, they must be catched on the singleflight group,
        // hence only 1

```

Reply:
I think this test is close, the fake client should just record the request count per namespace it was asked for, and you should make requests for a couple different namespaces to make sure singleflights share per namespace but not across them

Reply:
I tweaked it so that it should meet the test requirements @liggitt  @aojea